pl_bolts/models/detection/yolo/yolo_layers.py

```diff
@@ -0,0 +1,553 @@
+    """
+    half_wh = wh / 2
+    top_left = xy - half_wh
+    bottom_right = xy + half_wh
+    return torch.cat((top_left, bottom_right), -1)
+
+
+def _aligned_iou(dims1, dims2):
+    """
+    Calculates a matrix of intersections over union from box dimensions, assuming that the boxes
+    are located at the same coordinates.
+
+    Args:
+        dims1 (Tensor[N, 2]): width and height of N boxes
+        dims2 (Tensor[M, 2]): width and height of M boxes
+
+    Returns:
+        iou (Tensor[N, M]): the NxM matrix containing the pairwise IoU values for every element in
+            `dims1` and `dims2`
+    """
+    area1 = dims1[:, 0] * dims1[:, 1]  # [N]
+    area2 = dims2[:, 0] * dims2[:, 1]  # [M]
+
+    inter_wh = torch.min(dims1[:, None, :], dims2)  # [N, M, 2]
+    inter = inter_wh[:, :, 0] * inter_wh[:, :, 1]  # [N, M]
+    union = area1[:, None] + area2 - inter  # [N, M]
+
+    return inter / union
+
+
+def _elementwise_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
+    """
+    Returns the elementwise intersection-over-union between two sets of boxes.
+
+    Both sets of boxes are expected to be in (x1, y1, x2, y2) format.
+
+    Args:
+        boxes1 (Tensor[N, 4])
+        boxes2 (Tensor[N, 4])
+
+    Returns:
+        iou (Tensor[N]): the vector containing the elementwise IoU values for every element in
+        boxes1 and boxes2
+    """
+    area1 = box_area(boxes1)
+    area2 = box_area(boxes2)
+
+    lt = torch.max(boxes1[:, :2], boxes2[:, :2])  # [N,2]
+    rb = torch.min(boxes1[:, 2:], boxes2[:, 2:])  # [N,2]
+
+    wh = (rb - lt).clamp(min=0)  # [N,2]
+    inter = wh[:, 0] * wh[:, 1]  # [N]
+
+    iou = inter / (area1 + area2 - inter)
+    return iou
+
+
+def _elementwise_generalized_iou(boxes1: Tensor, boxes2: Tensor) -> Tensor:
+    """
+    Returns the elementwise generalized intersection-over-union between two sets of boxes.
+
+    Both sets of boxes are expected to be in (x1, y1, x2, y2) format.
+
+    Args:
+        boxes1 (Tensor[N, 4])
+        boxes2 (Tensor[N, 4])
+
+    Returns:
+        generalized_iou (Tensor[N]): the vector containing the elementwise generalized IoU values
+        for every element in boxes1 and boxes2
+    """
+
+    # Degenerate boxes give inf / nan results, so do an early check.
+    assert (boxes1[:, 2:] >= boxes1[:, :2]).all()
+    assert (boxes2[:, 2:] >= boxes2[:, :2]).all()
+
+    area1 = box_area(boxes1)
+    area2 = box_area(boxes2)
+
+    lt = torch.max(boxes1[:, :2], boxes2[:, :2])  # [N,2]
+    rb = torch.min(boxes1[:, 2:], boxes2[:, 2:])  # [N,2]
+
+    wh = (rb - lt).clamp(min=0)  # [N,2]
+    inter = wh[:, 0] * wh[:, 1]  # [N]
+
+    union = area1 + area2 - inter
+
+    iou = inter / union
+
+    lti = torch.min(boxes1[:, :2], boxes2[:, :2])
+    rbi = torch.max(boxes1[:, 2:], boxes2[:, 2:])
+
+    whi = (rbi - lti).clamp(min=0)  # [N,2]
+    areai = whi[:, 0] * whi[:, 1]
+
+    return iou - (areai - union) / areai
+
+
+class IoULoss(nn.Module):
```

Comment:
1 - `box_iou` simply ?

Reply:
To be as generic as possible, any loss function can be given to the detection layer as a parameter. So something like this would be possible: `lambda inputs, target: 1.0 - _elementwise_iou(inputs, target)`. Not that `lambda inputs, target: 1.0 - box_iou(inputs, target)` is a different thing. It calculates the loss between every (detector, target) pair, while YOLO matches the detectors to the targets and only calculates the loss from the matched pairs only.

Reply:
Same as https://github.com/PyTorchLightning/pytorch-lightning-bolts/pull/552#discussion_r575473956, right?